pkg/kubernetes/kubernetes.go

```diff
@@ -27,16 +29,35 @@ import (
 	"sigs.k8s.io/vsphere-csi-driver/pkg/csi/service/common"
 )
 
+var kubeconfig *string
+
+func init() {
+	kubeconfig = flag.String("kubeconfig", "", "path to the kubeconfig file")
+}
+
 // NewClient creates a newk8s client based on a service account
 func NewClient() (clientset.Interface, error) {
+	kubecfgPath := os.Getenv(clientcmd.RecommendedConfigPathEnvVar)
+	if *kubeconfig != "" {
+		kubecfgPath = *kubeconfig
+	}
 
 	var config *restclient.Config
 	var err error
-	klog.V(2).Info("k8s client using in-cluster config")
-	config, err = restclient.InClusterConfig()
-	if err != nil {
-		klog.Errorf("InClusterConfig failed %q", err)
-		return nil, err
+	if kubecfgPath != "" {
+		klog.V(2).Infof("k8s client using kubeconfig from %s", kubecfgPath)
+		config, err = clientcmd.BuildConfigFromFlags("", kubecfgPath)
+		if err != nil {
+			klog.Errorf("BuildConfigFromFlags failed %q", err)
+			return nil, err
+		}
+	} else {
```

Comment:
What about defaulting to an in cluster config if `BuildConfigFromFlags` fails?

Reply:
This would be very confusing. You specify a kubeconfig, but if something fails the controller suddenly tries to control the cluster it is running on.
Better to fail early and report it. If you want to use in-cluster config, just do not provide a kubeconfig.
BTW, your colleges of the vSphere cloud manager are using exactly the same logic, see https://github.com/kubernetes/cloud-provider-vsphere/blob/master/pkg/common/kubernetes/kubernetes.go#L38-L44